### PR TITLE
Unblock PRs because of protected jobs/contexts

### PR DESCRIPTION
**Pull request by owais**
This unblocks some PR jobs from not running right now.

------------------

We recently switch to CircleCI project environment variables to CircleCI Org Contexts. This allows us to restrict secrets like tokens to Docker Hub and Github to specific CI jobs and Github teams. Today the contexts can be used by any OpenTelemetry GitHub org member. By used, I mean any Otel member can push a commit and run any of the jobs that depend on the secrets provided by CircleCI contexts such as DockerHub (publish publish-dev) and GitHub (loadtest, windows-test).

While this adds much needed additional protection to secrets used by the project, it also means forked PRs from non-members will not be able to run the jobs that depend on any secrets such as loadtest and windows-test.

This PR re-structures the jobs a bit so that forked PRs don't need access to any secrets at all. This change ensures that only jobs triggered by commits to the main branch and release tags can run jobs that depend on secrets.

Merging this PR will unblock other PRs from non-members such as: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3422

This will l allow PR builds from non-opentelemetry members to pass and allow them to be merged but these builds will still fail once they're merged to main. I'll work with the maintainers to solve this for main branch builds.

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
@owais please add some comments to the circleci.yaml file to document what happens when. I have a hard time understanding what runs for PRs what runs for "main", what uses secrets and what doesn't. Generally we need to document the build files better, should treat it like a code in a language that is not very self-documenting :-)

**Comment by owais** [comment]
@tigrannajaryan added some comments to the CI file.

File: .circleci/config.yml
Comment (tigrannajaryan) [inline .circleci/config.yml:17]: This sentence is a bit confusing. Did it mean to refer to `loadtest-with-github-reports`? Is it an example of a job that uses this context?

File: .circleci/config.yml
Comment (tigrannajaryan) [inline .circleci/config.yml:22]: Should we further limit this to maintainers only or it would make impossible to make releases?

File: .circleci/config.yml
Comment (tigrannajaryan) [inline .circleci/config.yml:178]: Do we use _ or - as separator in job names?

**Review by tigrannajaryan — COMMENTED** [review]
Is there any way to break down CircleCI config.yml into smaller files and separate some of the concerns? I find it very hard to understand how the build works despite some very nice comments you added.

**Comment by tigrannajaryan** [comment]
> Is there any way to break down CircleCI config.yml into smaller files and separate some of the concerns? I find it very hard to understand how the build works despite some very nice comments you added.

Does not need to be this PR, but I wonder if this is possible at all. If not then maybe we refactor/rewrite the build config file to make it more understandable (not in this PR).

**Comment by tigrannajaryan** [comment]
Maybe we need a diagram or something that explains the workflows, jobs, steps, etc summarized somewhere so that we don't have to read the entire file to understand what is going on.

**Comment by gramidt** [comment]
This all makes sense and I agree with @tigrannajaryan on the diagram. 

I am curious how much time we'll want to invest in this given the eventual migration to GitHub Actions? I know we have some work to do before then, but a couple vendor distributions of the collector (including our F5 distribution), either have or are currently working on GitHub workflow pipelines, so I feel that there won't be too much heavy lifting when we're ready to migrate.

File: .circleci/config.yml
Comment (owais) [reply .circleci/config.yml:22]: It'll mean only maintainers will be able to release. We can create another group called `release-engineers` or something like that and add some maintainer and approver to it, then limit this context to that group.

BTW I need help setting this up as I'm not an admin (pinged you on slack).

File: .circleci/config.yml
Comment (owais) [reply .circleci/config.yml:178]: For commands, we use `_` but jobs use `-`. I tried to follow existing pattern.

File: .circleci/config.yml
Comment (owais) [reply .circleci/config.yml:17]: Yes. Fixed.

File: .circleci/config.yml
Comment (owais) [inline .circleci/config.yml:391]: Moved these steps to the re-usable `run_windows_test` command.

File: .circleci/config.yml
Comment (owais) [inline .circleci/config.yml:370]: Moved these steps to the re-usable `run_loadtest` command.

File: .circleci/config.yml
Comment (jrcamp) [inline .circleci/config.yml:17]: It's not clear to me why member of OT org would need access to this. I thought the loadtest that creates github issue only ran on main branch and not as part of any PRs? I don't think we've ever created GH issues as part of tests failing in a PR.

Comment (owais) [reply .circleci/config.yml:17]: >  I thought the loadtest that creates github issue only ran on main branch and not as part of any PRs? I don't think we've ever created GH issues as part of tests failing in a PR.

This is true. It only runs on main but now that it used Contexts to inject GITHUB_TOKEN into the job, we need to enable the context specifically for this job in the pipeline. As soon as we do that, the job can only be executed for commits pushed by people who are allowed to use the context which by default is all members of the OpenTelemetry Github org. So we now have two identical jobs for loadtest (and windows-test). One used contexts and run on main and the other does not use context and runs on PRs.  This unblocks PRs from the wider community.

**Comment by owais** [comment]
> Maybe we need a diagram or something that explains the workflows, jobs, steps, etc summarized somewhere so that we don't have to read the entire file to understand what is going on.
> This all makes sense and I agree with @tigrannajaryan on the diagram.

I don't know if I can do any better than CircleCI :) 


- This is what PR workflows looks like (taken from CircleCI):

image

- This is what master builds will look like

image


Release jobs would be the same as main jobs but will run `publish-stable` instead of `publish-check` as the last step.

**Comment by owais** [comment]
> I am curious how much time we'll want to invest in this given the eventual migration to GitHub Actions? I know we have some work to do before then, but a couple vendor distributions of the collector (including our F5 distribution), either have or are currently working on GitHub workflow pipelines, so I feel that there won't be too much heavy lifting when we're ready to migrate.

This makes sense to me. I think we should merge this to unblock PRs and then work towards moving to Github Actions and resolve any remaining issues there.

**Review by jrcamp — APPROVED** [review]

**Review by gramidt — APPROVED** [review]

**Pull request merged**
